#! /bin/sh /usr/share/dpatch/dpatch-run
## miscfixes.dpatch by Daniel Kristjansson <danielk@cuymedia.net>

@DPATCH@
diff -urNad mythtv-0.20+fixes14472~/docs/doxygen-architecture-document.cpp mythtv-0.20+fixes14472/docs/doxygen-architecture-document.cpp
--- blah/docs/doxygen-architecture-document.cpp	(revision 14472)
+++ blah/docs/doxygen-architecture-document.cpp	(working copy)
@@ -1,5 +1,5 @@
 /**
-\mainpage %MythTV Architecture
+\mainpage MythTV Architecture
 
 \section intro Introduction
 
@@ -59,7 +59,7 @@
       \ref av_player_subsystem "A/V players" are supported by libmythtv.
   <dt>libmythui              <dd>Main user interface rendering library
   <dt>libavcodec/libavformat <dd>This is the ffmpeg A/V decoding library (aka avlib).
-      <a href="http://ffmpeg.sourceforge.net/documentation.php">Documented Externally</a>.
+      <a href="http://ffmpeg.mplayerhq.hu/documentation.html">Documented Externally</a>.
   <dt>libmythmpeg2           <dd>Alternate MPEG-1/2 A/V decoding library.
       <a href="http://libmpeg2.sourceforge.net/">External Website</a>.
   <dt>libmythsamplerate      <dd>Audio resampling library
@@ -67,7 +67,7 @@
       We use this to support a different output sample rates than the sample
       rate used in the audio streams we play.
   <dt>libmythsoundtouch      <dd>Pitch preserving audio resampling library.
-      <a href="http://sky.prohosting.com/oparviai/soundtouch/">External Website</a>.
+      <a href="http://www.surina.net/soundtouch/">External Website</a>.
       We use this for the time-stretch feature.
 </dl>
 Two libraries libmythmpeg2 and libmythsamplerate appear redundant, but
diff -urNad mythtv-0.20+fixes14472~/docs/doxygen-create-developer-docs.cfg mythtv-0.20+fixes14472/docs/doxygen-create-developer-docs.cfg
--- blah/docs/doxygen-create-developer-docs.cfg	(revision 14472)
+++ blah/docs/doxygen-create-developer-docs.cfg	(working copy)
@@ -416,7 +416,7 @@
 # directories like "/usr/src/myproject". Separate the files or directories 
 # with spaces.
 
-INPUT                  = ../
+INPUT                  = doxygen-architecture-document.cpp ../filters ../libs ../programs ../../mythplugins
 
 # If the value of the INPUT tag contains directories, you can use the 
 # FILE_PATTERNS tag to specify one or more wildcard pattern (like *.cpp 
@@ -425,7 +425,7 @@
 # *.c *.cc *.cxx *.cpp *.c++ *.java *.ii *.ixx *.ipp *.i++ *.inl *.h *.hh *.hxx *.hpp 
 # *.h++ *.idl *.odl *.cs *.php *.php3 *.inc *.m *.mm
 
-FILE_PATTERNS          = *.cpp *.c *.h
+FILE_PATTERNS          = *.cpp *.c *.h *.mm
 
 # The RECURSIVE tag can be used to turn specify whether or not subdirectories 
 # should be searched for input files as well. Possible values are YES and NO. 
@@ -900,7 +900,7 @@
 # evaluate all C-preprocessor directives found in the sources and include 
 # files.
 
-ENABLE_PREPROCESSING   = YES
+ENABLE_PREPROCESSING   = NO
 
 # If the MACRO_EXPANSION tag is set to YES Doxygen will expand all macro 
 # names in the source code. If set to NO (the default) only conditional 
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/videoout_quartz.cpp mythtv-0.20+fixes14472/libs/libmythtv/videoout_quartz.cpp
--- blah/libs/libmythtv/videoout_quartz.cpp	(revision 14472)
+++ blah/libs/libmythtv/videoout_quartz.cpp	(working copy)
@@ -308,8 +308,13 @@
                                 .arg(name).arg(sw).arg(sh));
 
     // scale for non-square pixels
-    if (fabsf(aspect - (sw * 1.0 / sh)) > 0.01)
-    {
+    float realAspect = sw * 1.0 / sh; 
+    if (fabsf(aspect - realAspect) > 0.015) 
+    { 
+        VERBOSE(VB_PLAYBACK, QString("Image aspect doesn't match" 
+                                     " its resolution (%1 vs %2).") 
+                             .arg(aspect).arg(realAspect)); 
+
         if (parentData->scaleUpVideo)
         {
             // scale width up, leave height alone
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/osdtypes.cpp mythtv-0.20+fixes14472/libs/libmythtv/osdtypes.cpp
--- blah/libs/libmythtv/osdtypes.cpp	(revision 14472)
+++ blah/libs/libmythtv/osdtypes.cpp	(working copy)
@@ -644,12 +644,20 @@
 
 //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
 
-OSDType::OSDType(const QString &name)
+OSDType::OSDType(const QString &name) :
+    m_lock(true),
+    m_hidden(false),
+    m_name(QDeepCopy<QString>(name)),
+    m_parent(NULL)
 {
-    m_hidden = false;
-    m_name = name;
 }
 
+QString OSDType::Name(void)
+{
+    QMutexLocker locker(&m_lock);
+    return QDeepCopy<QString>(m_name);
+}
+
 OSDType::~OSDType()
 {
 }
@@ -658,71 +666,76 @@
 
 OSDTypeText::OSDTypeText(const QString &name, TTFFont *font, 
                          const QString &text, QRect displayrect,
-                         float wmult, float hmult)
-           : OSDType(name)
-{
-    m_message = text;
-    m_default_msg = text;
-    m_font = font;
+                         float wmult, float hmult) :
+    OSDType(name),
+    m_displaysize(displayrect),
+    m_screensize(displayrect),
+    m_unbiasedsize(unbias(m_screensize, wmult, hmult)),
+    m_message(QDeepCopy<QString>(text)),
+    m_default_msg(QDeepCopy<QString>(text)),
 
-    m_altfont = NULL;
-    
-    m_displaysize = displayrect;
-    m_screensize = displayrect;
-    m_multiline = false;
-    m_centered = false;
-    m_right = false;
-    m_usingalt = false;
+    m_font(font),
+    m_altfont(NULL),
 
-    m_selected  = false;
-    m_button    = false;
-    m_entrynum  = -1;
-    m_cursorpos = 0;
+    m_centered(false),
+    m_right(false),
 
-    m_scroller = false;
-    m_scrollx = m_scrolly = 0;
-    m_scrollinit = false;
+    m_multiline(false),
+    m_usingalt(false),
 
-    m_linespacing = 1.5;
+    m_selected(false),
+    m_button(false),
+    m_entrynum(-1),
+    m_cursorpos(0),
 
-    m_unbiasedsize = unbias(m_screensize, wmult, hmult);
+    m_scroller(false),
+    m_scrollx(0),
+    m_scrolly(0),
 
-    m_draw_info_str = "";
-    m_draw_info_len = 0;
+    m_scrollinit(false),
+
+    m_linespacing(1.5f),
+
+    m_draw_info_str(""),
+    m_draw_info_len(0)
+{
 }
 
-OSDTypeText::OSDTypeText(const OSDTypeText &other)
-           : OSDType(other.m_name)
+OSDTypeText::OSDTypeText(const OSDTypeText &other) :
+    OSDType(other.m_name),
+    m_draw_info_str(""),
+    m_draw_info_len(0)
 {
-    m_message = other.m_message;
-    m_default_msg = other.m_default_msg;
+    QMutexLocker locker(&other.m_lock);
+
+    m_displaysize = other.m_displaysize;
+    m_screensize = other.m_screensize;
+    m_unbiasedsize = other.m_unbiasedsize;
+
+    m_message = QDeepCopy<QString>(other.m_message);
+    m_default_msg = QDeepCopy<QString>(other.m_default_msg);
+
     m_font = other.m_font;
-
     m_altfont = other.m_altfont;
 
-    m_displaysize = other.m_displaysize;
-    m_screensize = other.m_screensize; 
-    m_multiline = other.m_multiline;
     m_centered = other.m_centered;
     m_right = other.m_right;
-    m_usingalt = other.m_usingalt; 
-   
-    m_selected = other.m_selected; 
-    m_button   = other.m_button; 
-    m_entrynum = other.m_entrynum; 
-    m_cursorpos = other.m_cursorpos; 
- 
+
+    m_multiline = other.m_multiline;
+    m_usingalt = other.m_usingalt;
+
+    m_selected = other.m_selected;
+    m_button   = other.m_button;
+    m_entrynum = other.m_entrynum;
+    m_cursorpos = other.m_cursorpos;
+
     m_scroller = other.m_scroller;
     m_scrollx = other.m_scrollx;
     m_scrolly = other.m_scrolly;
+
     m_scrollinit = other.m_scrollinit;
 
     m_linespacing = other.m_linespacing;
-
-    m_unbiasedsize = other.m_unbiasedsize;
-
-    m_draw_info_str = "";
-    m_draw_info_len = 0;
 }
 
 OSDTypeText::~OSDTypeText()
@@ -731,31 +744,86 @@
 
 void OSDTypeText::SetAltFont(TTFFont *font)
 {
+    QMutexLocker locker(&m_lock);
     m_altfont = font;
 }
 
 void OSDTypeText::SetText(const QString &text)
 {
-    m_message = text;
+    QMutexLocker locker(&m_lock);
+    m_message = QDeepCopy<QString>(text);
     m_cursorpos = text.length();
     m_scrollinit = false;
 }
 
+QString OSDTypeText::GetText(void) const
+{
+    QMutexLocker locker(&m_lock);
+    return QDeepCopy<QString>(m_message);
+}
+
 void OSDTypeText::SetDefaultText(const QString &text)
 {
-    m_message = text;
-    m_default_msg = text;
+    QMutexLocker locker(&m_lock);
+    m_message = QDeepCopy<QString>(text);
+    m_default_msg = QDeepCopy<QString>(text);
     m_scrollinit = false;
 }
+  
+QString OSDTypeText::GetDefaultText(void) const
+{
+    QMutexLocker locker(&m_lock);
+    return QDeepCopy<QString>(m_default_msg);
+}
 
+void OSDTypeText::SetMultiLine(bool multi)
+{
+    QMutexLocker locker(&m_lock);
+    m_multiline = multi;
+}
+
+void OSDTypeText::SetCentered(bool docenter)
+{
+    QMutexLocker locker(&m_lock);
+    m_centered = docenter;
+}
+
+void OSDTypeText::SetRightJustified(bool right)
+{
+    QMutexLocker locker(&m_lock);
+    m_right = right;
+}
+
+void OSDTypeText::SetScrolling(int x, int y)
+{
+    QMutexLocker locker(&m_lock);
+    m_scroller = true;
+    m_scrollx = x;
+    m_scrolly = y;
+}
+
+void OSDTypeText::SetLineSpacing(float linespacing)
+{
+    QMutexLocker locker(&m_lock);
+    m_linespacing = linespacing;
+}
+
 void OSDTypeText::Reinit(float wmult, float hmult)
 {
+    QMutexLocker locker(&m_lock);
     m_displaysize = m_screensize = bias(m_unbiasedsize, wmult, hmult);
 }
 
 void OSDTypeText::Draw(OSDSurface *surface, int fade, int maxfade, int xoff, 
                        int yoff)
 {
+    QMutexLocker locker(&m_lock);
+
+    // initialize regular expressions once, this is a very slow operation
+    static QMutex regexp_lock;
+    static QRegExp br("%BR%");
+    static QRegExp nl("\n");
+
     int textlength = 0;
 
     if (m_message == QString::null)
@@ -773,9 +841,11 @@
 
     if (m_multiline)
     {
-        QString tmp_msg = m_message;
-        tmp_msg.replace(QRegExp("%BR%"), "\n");
-        tmp_msg.replace(QRegExp("\n")," \n ");
+        QString tmp_msg = QDeepCopy<QString>(m_message);
+        regexp_lock.lock();
+        tmp_msg.replace(br, "\n");
+        tmp_msg.replace(nl," \n ");
+        regexp_lock.unlock();
 
         QStringList wordlist = QStringList::split(" ", tmp_msg);
         int length = 0;
@@ -902,6 +972,8 @@
                                    const QString &text, int fade, int maxfade, 
                                    int xoff, int yoff, bool double_size)
 {
+    QMutexLocker locker(&m_lock);
+
     if (m_draw_info_str != text)
     {
         m_draw_info_str = QDeepCopy<QString>(text);
@@ -1015,6 +1087,8 @@
                              const QString &text, int fade, int maxfade, 
                              int xoff, int yoff, bool doubl)
 {
+    QMutexLocker locker(&m_lock);
+
     if (m_centered || m_right)
     {
         int textlength = 0;
@@ -1067,6 +1141,8 @@
 
 bool OSDTypeText::MoveCursor(int dir)
 {
+    QMutexLocker locker(&m_lock);
+
     if (!IsEntry() || IsButton())
         return false;
 
@@ -1077,6 +1153,8 @@
 
 bool OSDTypeText::Delete(int dir)
 {
+    QMutexLocker locker(&m_lock);
+
     if (!IsEntry() || IsButton())
         return false;
 
@@ -1099,6 +1177,8 @@
 
 void OSDTypeText::InsertCharacter(QChar ch)
 {
+    QMutexLocker locker(&m_lock);
+
     if (!IsEntry() || IsButton())
         return;
 
Index: libs/libmythtv/osdtypes.h
===================================================================
--- blah/libs/libmythtv/osdtypes.h	(revision 14472)
+++ blah/libs/libmythtv/osdtypes.h	(working copy)
@@ -163,7 +163,7 @@
     void Hide(bool hidden = true) { m_hidden = hidden; }
     bool isHidden(void) { return m_hidden; }
 
-    QString Name() { return m_name; }
+    QString Name(void);
 
     virtual void Reinit(float wmult, float hmult) = 0;
 
@@ -171,6 +171,7 @@
                       int yoff) = 0;
 
   protected:
+    mutable QMutex m_lock;
     bool m_hidden;
     QString m_name;
     OSDSet *m_parent;
@@ -202,25 +203,24 @@
     void SetUseAlt(bool usealt) { m_usingalt = usealt; }
 
     void SetText(const QString &text);
-    QString GetText() { return m_message; }
+    QString GetText(void) const;
 
     void SetDefaultText(const QString &text);
-    QString GetDefaultText() { return m_default_msg; }
+    QString GetDefaultText(void) const;
 
-    void SetMultiLine(bool multi) { m_multiline = multi; }
-    bool GetMultiLine() { return m_multiline; }
+    void SetMultiLine(bool multi);
+    bool GetMultiLine(void) const { return m_multiline; }
 
-    void SetCentered(bool docenter) { m_centered = docenter; }
-    bool GetCentered() { return m_centered; }
+    void SetCentered(bool docenter);
+    bool GetCentered(void) const { return m_centered; }
 
-    void SetRightJustified(bool right) { m_right = right; }
-    bool GetRightJustified() { return m_right; }
+    void SetRightJustified(bool right);
+    bool GetRightJustified(void) const { return m_right; }
 
-    void SetScrolling(int x, int y) { m_scroller = true; m_scrollx = x;
-                                      m_scrolly = y; }
+    void SetScrolling(int x, int y);
 
-    void SetLineSpacing(float linespacing) { m_linespacing = linespacing; }
-    float GetLineSpacing() { return m_linespacing; }
+    void SetLineSpacing(float linespacing);
+    float GetLineSpacing(void) const { return m_linespacing; }
 
     void Draw(OSDSurface *surface, int fade, int maxfade, int xoff, int yoff);
     bool MoveCursor(int dir);
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/hdhrsignalmonitor.cpp mythtv-0.20+fixes14472/libs/libmythtv/hdhrsignalmonitor.cpp
--- blah/libs/libmythtv/hdhrsignalmonitor.cpp	(revision 14472)
+++ blah/libs/libmythtv/hdhrsignalmonitor.cpp	(working copy)
@@ -47,7 +47,7 @@
 
     _channel->DelAllPIDs();
 
-    signalStrength.SetThreshold(65);
+    signalStrength.SetThreshold(45);
 
     AddFlags(kDTVSigMon_WaitForSig);
 }
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/eithelper.cpp mythtv-0.20+fixes14472/libs/libmythtv/eithelper.cpp
--- blah/libs/libmythtv/eithelper.cpp	(revision 14472)
+++ blah/libs/libmythtv/eithelper.cpp	(working copy)
@@ -541,7 +541,6 @@
     fix[ 10001LL << 32 | 61441U << 16] = EITFixUp::kEFixPro7Sat;
     fix[ 10002LL << 32 | 61441U << 16] = EITFixUp::kEFixPro7Sat;
     fix[ 10003LL << 32 | 61441U << 16] = EITFixUp::kEFixPro7Sat;
-    fix[ 10005LL << 32 | 61441U << 16] = EITFixUp::kEFixPro7Sat;
     fix[ 10006LL << 32 | 61441U << 16] = EITFixUp::kEFixPro7Sat;
     fix[ 10009LL << 32 | 61441U << 16] = EITFixUp::kEFixPro7Sat;
     // On transport 10004 only DMAX needs no fixing:
@@ -563,27 +562,38 @@
         fix[10004LL<<32 | 61441U << 16 | 53618] = // K1010
         fix[10004LL<<32 | 61441U << 16 | 53619] = // GemsTV
         EITFixUp::kEFixPro7Sat;
+    // On transport 10005 QVC and Giga Digital  needs no fixing:
+    fix[    10005LL<<32 | 61441U << 16 | 50104] = // E! Entertainment
+        fix[10005LL<<32 | 61441U << 16 | 50107] = // 13th Street (KD)
+        fix[10005LL<<32 | 61441U << 16 | 50301] = // ESPN Classic
+        fix[10005LL<<32 | 61441U << 16 | 50302] = // VH1 Classic
+        fix[10005LL<<32 | 61441U << 16 | 50303] = // Wein TV
+        fix[10005LL<<32 | 61441U << 16 | 50304] = // AXN
+        fix[10005LL<<32 | 61441U << 16 | 50305] = // Silverline
+        fix[10005LL<<32 | 61441U << 16 | 50306] = // NASN
+        fix[10005LL<<32 | 61441U << 16 | 50307] = // Disney Toon
+        fix[10005LL<<32 | 61441U << 16 | 53105] = // NASN (engl)
+        fix[10005LL<<32 | 61441U << 16 | 53115] = // VH1 Classic (engl)
+        fix[10005LL<<32 | 61441U << 16 | 53405] = // ESPN Classic (engl)
+        fix[10005LL<<32 | 61441U << 16 | 53402] = // AXN (engl)
+        fix[10005LL<<32 | 61441U << 16 | 53613] = // CNN (engl)
+        fix[10005LL<<32 | 61441U << 16 | 53516] = // Voyages Television
+        fix[10005LL<<32 | 61441U << 16 | 53611] = // Der Schmuckkanal
+        fix[10005LL<<32 | 61441U << 16 | 53104] = // Jukebox
+        EITFixUp::kEFixPro7Sat;
     // On transport 10007 only following channels need fixing:
-    fix[    10007LL<<32| 61441U << 16 | 53605] = // terranova
-        fix[10007LL<<32| 61441U << 16 | 53607] = // Eurosport
+    fix[    10007LL<<32| 61441U << 16 | 53607] = // Eurosport
         fix[10007LL<<32| 61441U << 16 | 53608] = // Das Vierte
         fix[10007LL<<32| 61441U << 16 | 53609] = // Viva
         fix[10007LL<<32| 61441U << 16 | 53628] = // COMEDY CENTRAL
         EITFixUp::kEFixPro7Sat;
     // On transport 10008 only following channels need fixing:
     fix[    10008LL<<32 | 61441U << 16 | 53002] = // Tele 5
-        fix[10008LL<<32 | 61441U << 16 | 53624] = // DSF
         fix[10008LL<<32 | 61441U << 16 | 53630] = // HSE24
         EITFixUp::kEFixPro7Sat;
 
-    fix[ 774LL << 32 | 8468 << 16 | 16392] = EITFixUp::kEFixPro7Sat; //DVB-T Berlin dsf
     fix[ 772LL << 32 | 8468 << 16 | 16387] = EITFixUp::kEFixPro7Sat; //DVB-T Berlin HSE/MonA TV
     fix[8707LL << 32 | 8468 << 16 | 16413] = EITFixUp::kEFixPro7Sat; //DVB-T Ruhrgebiet Tele 5
-    fix[1082LL << 32 |    1 << 16 | 20001] = EITFixUp::kEFixPro7Sat; //DVB-S Pro7 Swiss
-    fix[1082LL << 32 |    1 << 16 | 20002] = EITFixUp::kEFixPro7Sat; //DVB-S Pro7 Austria
-    fix[1082LL << 32 |    1 << 16 | 20003] = EITFixUp::kEFixPro7Sat; //DVB-S Kabel1 Swiss
-    fix[1082LL << 32 |    1 << 16 | 20004] = EITFixUp::kEFixPro7Sat; //DVB-S Kabel1 Austria
-    fix[1082LL << 32 |    1 << 16 | 20005] = EITFixUp::kEFixPro7Sat; //DVB-S Sat.1 Austria
 
     fix[ 133 << 16] = EITFixUp::kEFixPro7Sat; // Premiere and Pro7/Sat.1
 }
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/datadirect.h mythtv-0.20+fixes14472/libs/libmythtv/datadirect.h
--- blah/libs/libmythtv/datadirect.h	(revision 14472)
+++ blah/libs/libmythtv/datadirect.h	(working copy)
@@ -215,7 +215,7 @@
 {
   public:
     DDStructureParser(DataDirectProcessor& _ddparent) :
-        parent(_ddparent), saw_repeat(false), saw_new(false) {}
+        parent(_ddparent) {}
 
     bool startElement(const QString &pnamespaceuri, const QString &plocalname,
                       const QString &pqname, const QXmlAttributes &pxmlatts);
@@ -240,8 +240,6 @@
     DataDirectProductionCrew curr_productioncrew;
     DataDirectGenre          curr_genre;
     QString                  lastprogramid;
-    bool                     saw_repeat;
-    bool                     saw_new;
 };
 
 
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/datadirect.cpp mythtv-0.20+fixes14472/libs/libmythtv/datadirect.cpp
--- blah/libs/libmythtv/datadirect.cpp	(revision 14472)
+++ blah/libs/libmythtv/datadirect.cpp	(working copy)
@@ -168,14 +168,8 @@
         curr_schedule.duration = QTime(durstr.mid(2, 2).toInt(), 
                                        durstr.mid(5, 2).toInt(), 0, 0);
 
-        QString isrepeat = pxmlatts.value("repeat");
-        curr_schedule.repeat = (isrepeat == "true");
-        saw_repeat |= !isrepeat.isEmpty();
-
-        QString isnew = pxmlatts.value("new");
-        curr_schedule.isnew = (isnew == "true");
-        saw_new |= !isnew.isEmpty();
-
+        curr_schedule.repeat = (pxmlatts.value("repeat") == "true");
+        curr_schedule.isnew = (pxmlatts.value("new") == "true");
         curr_schedule.stereo = (pxmlatts.value("stereo") == "true");
         curr_schedule.subtitled = (pxmlatts.value("subtitled") == "true");
         curr_schedule.hdtv = (pxmlatts.value("hdtv") == "true");
@@ -420,16 +414,6 @@
 
 bool DDStructureParser::endDocument() 
 {
-    MSqlQuery query(MSqlQuery::DDCon());
-    query.prepare(
-        "INSERT INTO dd_state (sawrepeat, sawnew) "
-        "VALUES (:SAWREPEAT, :SAWNEW)");
-    query.bindValue(":SAWREPEAT", saw_repeat);
-    query.bindValue(":SAWNEW",    saw_new);
-
-    if (!query.exec())
-        MythContext::DBError("Inserting into dd_state", query);
-
     return true;
 }
  
@@ -617,28 +601,6 @@
 {
     MSqlQuery query(MSqlQuery::DDCon());
 
-    query.prepare(
-        "SELECT sawrepeat, sawnew "
-        "FROM dd_state");
-
-    if (!query.exec())
-    {
-        MythContext::DBError("Querying into dd_state", query);
-        return;
-    }
-
-    if (!query.next())
-    {
-        VERBOSE(VB_IMPORTANT, LOC_ERR + "UpdateProgramViewTable no dd_state!");
-        return;
-    }
-
-    bool saw_repeat = query.value(0).toUInt(); (void) saw_repeat;
-    bool saw_new    = query.value(1).toUInt(); (void) saw_new;
-
-//    VERBOSE(VB_GENERAL, LOC + QString("Saw new: %1, saw repeat: %2")
-//            .arg(saw_new).arg(saw_repeat));
-
     if (!query.exec("TRUNCATE TABLE dd_v_program;"))
         MythContext::DBError("Truncating temporary table dd_v_program", query);
 
@@ -654,7 +616,7 @@
         "       tvrating,       mpaarating,      programid )      "
         "SELECT chanid,         scheduletime,    endtime,         "
         "       title,          subtitle,        description,     "
-        "       year,           stars,           %1,              "
+        "       year,           stars,           isrepeat,        "
         "       stereo,         subtitled,       hdtv,            "
         "       closecaptioned, partnumber,      parttotal,       "
         "       seriesid,       originalairdate, showtype,        "
@@ -665,13 +627,8 @@
         "       (channel.xmltvid       = dd_schedule.stationid) AND "
         "       (channel.sourceid      = :SOURCEID))";
 
-    QString repeat = (saw_new) ?
-        "(not isnew) AND "
-        "(SUBSTRING(dd_program.programid,1,2) IN ('EP', 'SH'))" :
-        "isrepeat";
+    query.prepare(qstr);
 
-    query.prepare(qstr.arg(repeat));
-
     query.bindValue(":SOURCEID", sourceid);
 
     if (!query.exec())
@@ -1112,11 +1069,11 @@
 
     if (cachedata)
     {
-        cache_dd_data = tmpDir + QString("/mythtv_dd_cache_%1_%2_%3_%4")
+        cache_dd_data = tmpDir + QString("/mythtv_dd_cache_%1_%2_UTC_%3_to_%4")
             .arg(GetListingsProvider())
             .arg(GetUserID().ascii())
-            .arg(pstartDate.toString())
-            .arg(pendDate.toString());
+            .arg(pstartDate.toString("yyyyMMddhhmmss"))
+            .arg(pendDate.toString("yyyyMMddhhmmss"));
 
         if (QFile(cache_dd_data).exists() && inputfilename.isEmpty())
         {
@@ -1244,9 +1201,6 @@
 {
     QMap<QString,QString> dd_tables;
 
-    dd_tables["dd_state"] =
-        "( sawrepeat bool,               sawnew bool )";
-
     dd_tables["dd_station"] =
         "( stationid char(12),           callsign char(10),     "
         "  stationname varchar(40),      affiliate varchar(25), "
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/atscstreamdata.cpp mythtv-0.20+fixes14472/libs/libmythtv/atscstreamdata.cpp
--- blah/libs/libmythtv/mpeg/atscstreamdata.cpp	(revision 14472)
+++ blah/libs/libmythtv/mpeg/atscstreamdata.cpp	(working copy)
@@ -461,7 +461,7 @@
     _sourceid_to_atsc_maj_min.clear();
     for (uint i = 0; i < vct->ChannelCount() ; i++)
     {
-        if (vct->IsHiddenInGuide(i))
+        if (vct->IsHidden(i) && vct->IsHiddenInGuide(i))
         {
             VERBOSE(VB_EIT, QString("%1 chan %2-%3 is hidden in guide")
                     .arg(vct->ModulationMode(i) == 1 ? "NTSC" : "ATSC")
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/vsync.cpp mythtv-0.20+fixes14472/libs/libmythtv/vsync.cpp
--- blah/libs/libmythtv/vsync.cpp	(revision 14472)
+++ blah/libs/libmythtv/vsync.cpp	(working copy)
@@ -444,9 +444,9 @@
     OpenGLVideoSyncPrivate()
     {
         m_glXGetVideoSyncSGI = (PFNGLXGETVIDEOSYNCSGIPROC)
-                glXGetProcAddress("glXGetVideoSyncSGI");
+            GetProcAddress("glXGetVideoSyncSGI");
         m_glXWaitVideoSyncSGI = (PFNGLXWAITVIDEOSYNCSGIPROC)
-                glXGetProcAddress("glXWaitVideoSyncSGI");
+            GetProcAddress("glXWaitVideoSyncSGI");
     }
 
     bool funcsLoaded()
@@ -466,10 +466,18 @@
     }
 
   private:
-    __GLXextFuncPtr glXGetProcAddress(const char * const procName)
+    __GLXextFuncPtr GetProcAddress(const char * const procName)
     {
-        __GLXextFuncPtr ret = glXGetProcAddressARB((const GLubyte *) procName);
+        __GLXextFuncPtr ret = NULL;
 
+#if GLX_VERSION_1_4
+        X11S(ret = glXGetProcAddress((const GLubyte*)procName));
+#elif GLX_ARB_get_proc_address
+        X11S(ret = glXGetProcAddressARB((const GLubyte*)procName));
+#elif GLX_EXT_get_proc_address
+        X11S(ret = glXGetProcAddressEXT((const GLubyte*)procName));
+#endif
+
         if (!ret)
         {
             VERBOSE(VB_PLAYBACK,
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/osdtypeteletext.cpp mythtv-0.20+fixes14472/libs/libmythtv/osdtypeteletext.cpp
--- blah/libs/libmythtv/osdtypeteletext.cpp	(revision 14472)
+++ blah/libs/libmythtv/osdtypeteletext.cpp	(working copy)
@@ -953,7 +953,7 @@
                 endbox = true;
                 goto ctrl;
             case 0x0b: // start box
-                if (x < kTeletextColumns - 1 && (page[x + 1] & 0x7F != 0x0b))
+                if (x < kTeletextColumns - 1 && ((page[x + 1] & 0x7F) != 0x0b))
                     startbox = true;
                 goto ctrl;
             case 0x0c: // normal height
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/tv_play.cpp mythtv-0.20+fixes14472/libs/libmythtv/tv_play.cpp
--- blah/libs/libmythtv/tv_play.cpp	(revision 14472)
+++ blah/libs/libmythtv/tv_play.cpp	(working copy)
@@ -55,6 +55,7 @@
 #define DEBUG_CHANNEL_PREFIX 0 /**< set to 1 to channel prefixing */
 #define DEBUG_ACTIONS 0 /**< set to 1 to debug actions */
 #define LOC QString("TV: ")
+#define LOC_WARN QString("TV Warning: ")
 #define LOC_ERR QString("TV Error: ")
 
 void TV::InitKeys(void)
@@ -1480,8 +1481,6 @@
 void TV::RunTV(void)
 { 
     paused = false;
-    QKeyEvent *keypressed;
-    QString netCmd;
 
     doing_ff_rew = 0;
     ff_rew_index = kInitFFRWSpeed;
@@ -1570,29 +1569,42 @@
             getRecorderPlaybackInfo = false;
         }
 
-        if (nvp && (keyList.count() > 0))
-        { 
+        bool had_key = false;
+        if (nvp)
+        {
+            QKeyEvent *keypressed = NULL;
+
             keyListLock.lock();
-            keypressed = keyList.first();
-            keyList.removeFirst();
+            if (keyList.count() > 0)
+            {
+                keypressed = keyList.first();
+                keyList.removeFirst();
+            }
             keyListLock.unlock();
 
-            ProcessKeypress(keypressed);
-            delete keypressed;
+            if (keypressed)
+            {
+                had_key = true;
+                ProcessKeypress(keypressed);
+                delete keypressed;
+            }
         }
 
-        netCmd = "";
-        ncLock.lock();
-        if (networkControlCommands.size())
+        if (nvp && !had_key && !ignoreKeys)
         {
-            netCmd = networkControlCommands.front();
-            networkControlCommands.pop_front();
+            QString netCmd = QString::null;
+            ncLock.lock();
+            if (networkControlCommands.size())
+            {
+                netCmd = networkControlCommands.front();
+                networkControlCommands.pop_front();
+            }
+            ncLock.unlock();
+
+            if (!netCmd.isEmpty())
+                processNetworkControlCommand(netCmd);
         }
-        ncLock.unlock();
 
-        if (netCmd != "")
-            processNetworkControlCommand(netCmd);
-
         if ((recorder && recorder->GetErrorStatus()) ||
             (nvp && nvp->IsErrored()) || IsErrored())
         {
@@ -2767,23 +2779,50 @@
 
 void TV::processNetworkControlCommand(QString command)
 {
+#ifdef DEBUG_ACTIONS
+    VERBOSE(VB_IMPORTANT, LOC + "ProcessNetworkControlCommand(" +
+            QString("%1)").arg(command));
+#endif
+
     QStringList tokens = QStringList::split(" ", command);
+    if (tokens.size() < 2)
+    {
+        VERBOSE(VB_IMPORTANT, LOC_ERR + "Not enough tokens"
+                "in network control command" + "\n\t\t\t" +
+                QString("'%1'").arg(command));
+        return;
+    }
 
+    if (!dialogname.isEmpty())
+    {
+        VERBOSE(VB_IMPORTANT, LOC_WARN + "Ignoring network "
+                "control command\n\t\t\t" +
+                QString("because dialog '%1'").arg(dialogname) +
+                "is waiting for a response");
+        return;
+    }
+
     if (tokens[1] != "QUERY")
         ClearOSD();
 
     if (tokens.size() == 3 && tokens[1] == "CHANID")
     {
-        ChangeChannel(tokens[2].toUInt(), "");
+        queuedChanID = tokens[2].toUInt();
+        queuedChanNum = QString::null;
+        CommitQueuedInput();
     }
     else if (tokens.size() == 3 && tokens[1] == "CHANNEL")
     {
-        if (tokens[2] == "UP")
-            ChangeChannel(CHANNEL_DIRECTION_UP);
-        else if (tokens[2] == "DOWN")
-            ChangeChannel(CHANNEL_DIRECTION_DOWN);
-        else if (tokens[2].contains(QRegExp("^[-\\.\\d_#]+$")))
-            ChangeChannel(0, tokens[2]);
+        uint aindx = (activenvp == nvp) ? 0 : 1;
+        if (StateIsLiveTV(GetState()) && !pseudoLiveTVState[aindx])
+        {
+            if (tokens[2] == "UP")
+                ChangeChannel(CHANNEL_DIRECTION_UP);
+            else if (tokens[2] == "DOWN")
+                ChangeChannel(CHANNEL_DIRECTION_DOWN);
+            else if (tokens[2].contains(QRegExp("^[-\\.\\d_#]+$")))
+                ChangeChannel(0, tokens[2]);
+        }
     }
     else if (tokens.size() == 3 && tokens[1] == "SPEED")
     {
@@ -2865,6 +2904,9 @@
         }
         else if (tokens[2].contains(QRegExp("^\\d+\\/\\d+x$")))
         {
+            if (activerbuffer->InDVDMenuOrStillFrame())
+                return;
+
             if (paused)
                 DoPause();
 
@@ -2904,6 +2946,9 @@
     }
     else if (tokens.size() >= 3 && tokens[1] == "SEEK" && activenvp)
     {
+        if (activerbuffer->InDVDMenuOrStillFrame())
+            return;
+
         if (tokens[2] == "BEGINNING")
             DoSeek(-activenvp->GetFramesPlayed(), tr("Jump to Beginning"));
         else if (tokens[2] == "FORWARD")
@@ -5198,7 +5243,7 @@
             if ((tokens[1] != "ANSWER") && (tokens[1] != "RESPONSE"))
             {
                 ncLock.lock();
-                networkControlCommands.push_back(message);
+                networkControlCommands.push_back(QDeepCopy<QString>(message));
                 ncLock.unlock();
             }
         }
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/videosource.cpp mythtv-0.20+fixes14472/libs/libmythtv/videosource.cpp
--- blah/libs/libmythtv/videosource.cpp	(revision 14472)
+++ blah/libs/libmythtv/videosource.cpp	(working copy)
@@ -24,6 +24,7 @@
 // MythTV headers
 #include "mythconfig.h"
 #include "mythwidgets.h"
+#include "mythdialogs.h"
 #include "mythcontext.h"
 #include "mythdbcon.h"
 #include "videosource.h"
@@ -35,6 +36,7 @@
 #include "frequencies.h"
 #include "diseqcsettings.h"
 
+
 #ifdef USING_DVB
 #include <linux/dvb/frontend.h>
 #endif
@@ -388,6 +390,11 @@
     if ( find_grabber_proc.start() )
     {
 
+        VERBOSE(VB_IMPORTANT, "Running tv_find_grabbers.");
+        MythBusyDialog *find_grabbers_dialog = new MythBusyDialog(
+            QObject::tr("Searching for installed XMLTV grabbers"));
+        find_grabbers_dialog->start();
+
         int i=0;
         // Assume it shouldn't take more than 25 seconds
         // Broken versions of QT cause QProcess::start
@@ -397,6 +404,8 @@
         {
             usleep(100000);
             ++i;
+            // Update the progress dialog without using the event loop
+            qApp->processEvents();
         }
 
         if (find_grabber_proc.normalExit())
@@ -417,6 +426,9 @@
         {
             VERBOSE(VB_IMPORTANT, "tv_find_grabbers exited early or we timed out waiting");
         }
+
+        find_grabbers_dialog->Close();
+        delete find_grabbers_dialog;
     }
     else
     {
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/videoout_xv.cpp mythtv-0.20+fixes14472/libs/libmythtv/videoout_xv.cpp
--- blah/libs/libmythtv/videoout_xv.cpp	(revision 14472)
+++ blah/libs/libmythtv/videoout_xv.cpp	(working copy)
@@ -418,11 +418,7 @@
             .arg(w).arg(h).arg(window_w).arg(window_h));
 
     // Determine if we are using Xinerama
-    int event_base, error_base;
-    bool usingXinerama = false;
-    X11S(usingXinerama = 
-         (XineramaQueryExtension(XJ_disp, &event_base, &error_base) &&
-          XineramaIsActive(XJ_disp)));
+    bool usingXinerama = (GetNumberOfXineramaScreens() > 1);
 
     // If the dimensions are invalid, assume square pixels and 17" screen.
     // Only print warning if this isn't Xinerama, we will fix Xinerama later.
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/tv_rec.cpp mythtv-0.20+fixes14472/libs/libmythtv/tv_rec.cpp
--- blah/libs/libmythtv/tv_rec.cpp	(revision 14472)
+++ blah/libs/libmythtv/tv_rec.cpp	(working copy)
@@ -2057,8 +2057,9 @@
 /** \fn QString add_spacer(const QString&, const QString&)
  *  \brief Adds the spacer before the last character in chan.
  */
-static QString add_spacer(const QString &chan, const QString &spacer)
+static QString add_spacer(const QString &channel, const QString &spacer)
 {
+    QString chan = QDeepCopy<QString>(channel);
     if ((chan.length() >= 2) && !spacer.isEmpty())
         return chan.left(chan.length()-1) + spacer + chan.right(1);
     return chan;
@@ -2096,6 +2097,10 @@
                                bool          &is_extra_char_useful,
                                QString       &needed_spacer)
 {
+#if DEBUG_CHANNEL_PREFIX
+    VERBOSE(VB_IMPORTANT, QString("CheckChannelPrefix(%1)").arg(prefix));
+#endif
+
     static const uint kSpacerListSize = 5;
     static const char* spacers[kSpacerListSize] = { "", "_", "-", "#", "." };
 
@@ -2122,7 +2127,8 @@
     {
         for (uint j = 0; j < kSpacerListSize; j++)
         {
-            QString qprefix = add_spacer(prefix, spacers[j]);
+	    QString qprefix = add_spacer(
+                prefix, (spacers[j] == "_") ? "\\_" : spacers[j]);
             query.prepare(basequery.arg(qprefix) + cardquery[i]);
 
             if (!query.exec() || !query.isActive())
@@ -2172,21 +2178,16 @@
     // sharing the prefix we were given.
 
     // Is an extra characher useful for disambiguation?
-    for (uint i = 0; i < fchannum.size(); i++)
-    {
-        if (fchannum[i] != add_spacer(prefix, fspacer[0]))
-        {
-            is_extra_char_useful = true;
-            break;
-        }
-    }
+    is_extra_char_useful = false;
+    for (uint i = 0; (i < fchannum.size()) && !is_extra_char_useful; i++)
+        is_extra_char_useful = (fchannum[i] != add_spacer(prefix, fspacer[i]));
 
     // Are any of the channels complete w/o spacer?
     // If so set is_complete_valid_channel_on_rec,
     // with a preference for our cardid.
     for (uint i = 0; i < fchannum.size(); i++)
     {
-        if (fchannum[i] == add_spacer(prefix, fspacer[0]))
+        if (fchannum[i] == prefix)
         {
             is_complete_valid_channel_on_rec = fcardid[i];
             if (fcardid[i] == (uint)cardid)
@@ -2194,20 +2195,26 @@
         }
     }
 
+    if (is_complete_valid_channel_on_rec)
+        return true;
+
     // Add a spacer, if one is needed to select a valid channel.
-    if (!is_complete_valid_channel_on_rec)
+    bool spacer_needed = true;
+    for (uint i = 0; (i < fspacer.size() && spacer_needed); i++)
+        spacer_needed = !fspacer[i].isEmpty();
+    if (spacer_needed)
+        needed_spacer = QDeepCopy<QString>(fspacer[0]);
+
+    // If it isn't useful to wait for more characters,
+    // then try to commit to any true match immediately.
+    for (uint i = 0; i < ((is_extra_char_useful) ? 0 : fchanid.size()); i++)
     {
-        bool spacer_needed = true;
-        for (uint i = 0; i < fspacer.size(); i++)
+        if (fchannum[i] == add_spacer(prefix, fspacer[i]))
         {
-            if (fspacer[i].isEmpty())
-            {
-                spacer_needed = false;
-                break;
-            }
+            needed_spacer = QDeepCopy<QString>(fspacer[i]);
+            is_complete_valid_channel_on_rec = fcardid[i];
+            return true;
         }
-        if (spacer_needed)
-            needed_spacer = QDeepCopy<QString>(fspacer[0]);
     }
 
     return true;
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/programinfo.cpp mythtv-0.20+fixes14472/libs/libmythtv/programinfo.cpp
--- blah/libs/libmythtv/programinfo.cpp	(revision 14472)
+++ blah/libs/libmythtv/programinfo.cpp	(working copy)
@@ -3172,7 +3172,7 @@
     if (record == NULL && recordid)
     {
         record = new ScheduledRecording();
-        record->loadByID(recordid);
+        record->loadByProgram(this);
     }
 
     if (filesize > 0)
diff -urNad mythtv-0.20+fixes14472~/libs/libmythtv/cpu_accel.c mythtv-0.20+fixes14472/libs/libmythtv/cpu_accel.c
--- blah/libs/libmythmpeg2/cpu_accel.c	(revision 14472)
+++ blah/libs/libmythmpeg2/cpu_accel.c	(working copy)
@@ -190,7 +190,7 @@
 	accel |= MPEG2_ACCEL_SPARC_VIS;
 
 #ifdef ACCEL_DETECT
-    if (accel & (MPEG2_ACCEL_SPARC_VIS2 | MPEG2_ACCEL_DETECT) ==
+    if ((accel & (MPEG2_ACCEL_SPARC_VIS2 | MPEG2_ACCEL_DETECT)) ==
 	MPEG2_ACCEL_DETECT) {
 	static RETSIGTYPE (* oldsig) (int);
 
diff -urNad mythtv-0.20+fixes14472~/libs/libmyth/mythdialogs.cpp mythtv-0.20+fixes14472/libs/libmyth/mythdialogs.cpp
--- blah/libs/libmyth/mythdialogs.cpp	(revision 14472)
+++ blah/libs/libmyth/mythdialogs.cpp	(working copy)
@@ -471,24 +471,65 @@
 void MythPopupBox::defaultButtonPressedHandler(void)
 {
     const QObjectList *objlist = children();
-    QObjectListIt it(*objlist);
+    QObjectListIt itf(*objlist);
     QObject *objs;
     int i = 0;
-    while ((objs = it.current()) != 0)
+    bool foundbutton = false;
+
+    // this bit of code will work if the window is focused
+    while ((objs = itf.current()) != 0)
     {
-        ++it;
+        ++itf;
         if (objs->isWidgetType())
         {
             QWidget *widget = (QWidget *)objs;
             if (widget->isA("MythPushButton"))
             {
                 if (widget->hasFocus())
+                {
+                    foundbutton = true;
                     break;
+                }
                 i++;
             }
         }
     }
-    done(i);
+    if (foundbutton)
+    {
+        done(i);
+        return;
+    }
+
+    // this bit of code should always work but requires a cast
+    QObjectListIt itd(*objlist);
+    i = 0;
+    while ((objs = itd.current()) != 0)
+    {
+        ++itd;
+        if (objs->isWidgetType())
+        {
+            QWidget *widget = (QWidget *)objs;
+            if (widget->isA("MythPushButton"))
+            {
+                MythPushButton *button = dynamic_cast<MythPushButton*>(widget);
+                if (button && button->isDown())
+                {
+                    foundbutton = true;
+                    break;
+                }
+                i++;
+            }
+        }
+    }
+    if (foundbutton)
+    {
+        done(i);
+        return;
+    }
+
+    VERBOSE(VB_IMPORTANT, "MythPopupBox::defaultButtonPressedHandler(void)"
+            "\n\t\t\tWe should never get here!");
+    done(Rejected);
 }
 
 void MythPopupBox::defaultExitHandler()
@@ -550,39 +591,49 @@
                                    QString message, QString button1msg,
                                    QString button2msg, int defvalue)
 {
-    MythPopupBox popup(parent, title);
+    MythPopupBox *popup = new MythPopupBox(parent, title);
 
-    popup.addLabel(message, Medium, true);
-    popup.addLabel("");
+    popup->addLabel(message, Medium, true);
+    popup->addLabel("");
 
-    QButton *but1 = popup.addButton(button1msg);
-    QButton *but2 = popup.addButton(button2msg);
+    QButton *but1 = popup->addButton(button1msg);
+    QButton *but2 = popup->addButton(button2msg);
 
     if (defvalue == 1)
         but1->setFocus();
     else
         but2->setFocus();
 
-    return popup.ExecPopup();
+    int ret = popup->ExecPopup();
+
+    popup->hide();
+    popup->deleteLater();
+
+    return ret;
 }
 
 int MythPopupBox::showButtonPopup(MythMainWindow *parent, QString title,
                                   QString message, QStringList buttonmsgs,
                                   int defvalue)
 {
-    MythPopupBox popup(parent, title);
+    MythPopupBox *popup = new MythPopupBox(parent, title);
 
-    popup.addLabel(message, Medium, true);
-    popup.addLabel("");
+    popup->addLabel(message, Medium, true);
+    popup->addLabel("");
 
     for (unsigned int i = 0; i < buttonmsgs.size(); i++ )
     {
-        QButton *but = popup.addButton(buttonmsgs[i]);
+        QButton *but = popup->addButton(buttonmsgs[i]);
         if (defvalue == (int)i)
             but->setFocus();
     }
 
-    return popup.ExecPopup();
+    int ret = popup->ExecPopup();
+
+    popup->hide();
+    popup->deleteLater();
+
+    return ret;
 }
 
 MythProgressDialog::MythProgressDialog(const QString &message, int totalSteps)
diff -urNad mythtv-0.20+fixes14472~/libs/libmyth/mythsocket.cpp mythtv-0.20+fixes14472/libs/libmyth/mythsocket.cpp
--- blah/libs/libmyth/mythsocket.cpp	(revision 14472)
+++ blah/libs/libmyth/mythsocket.cpp	(working copy)
@@ -689,7 +689,6 @@
 {
     VERBOSE(VB_SOCKET, "MythSocket: readyread thread start");
     fd_set rfds;
-    struct timeval timeout;
     MythSocket *sock;
     int maxfd;
     bool found;
@@ -737,10 +736,7 @@
             ++it;
         }
 
-        timeout.tv_sec = 0;
-        timeout.tv_usec = 10000;
-
-        int rval = select(maxfd + 1, &rfds, NULL, NULL, &timeout);
+        int rval = select(maxfd + 1, &rfds, NULL, NULL, NULL);
         if (rval == -1)
         {
             VERBOSE(VB_SOCKET, "MythSocket: select returned error");
diff -urNad mythtv-0.20+fixes14472~/libs/libavcodec/libavcodec.pro mythtv-0.20+fixes14472/libs/libavcodec/libavcodec.pro
--- blah/libs/libavcodec/libavcodec.pro	(revision 14472)
+++ blah/libs/libavcodec/libavcodec.pro	(working copy)
@@ -17,7 +17,7 @@
     QMAKE_CFLAGS_SHLIB = 
 }
 
-QMAKE_CFLAGS_DEBUG += -O
+!profile:QMAKE_CFLAGS_DEBUG += -O
 
 QMAKE_CLEAN += $(TARGET) $(TARGETA) $(TARGETD) $(TARGET0) $(TARGET1) $(TARGET2)
 
diff -urNad mythtv-0.20+fixes14472~/libs/libavutil/libavutil.pro mythtv-0.20+fixes14472/libs/libavutil/libavutil.pro
--- blah/libs/libavutil/libavutil.pro	(revision 14472)
+++ blah/libs/libavutil/libavutil.pro	(working copy)
@@ -18,7 +18,7 @@
     QMAKE_CFLAGS_SHLIB = 
 }
 
-QMAKE_CFLAGS_DEBUG += -O
+!profile:QMAKE_CFLAGS_DEBUG += -O
 
 QMAKE_CLEAN += $(TARGET) $(TARGETA) $(TARGETD) $(TARGET0) $(TARGET1) $(TARGET2)
 
diff -urNad mythtv-0.20+fixes14472~/programs/mythfrontend/globalsettings.cpp mythtv-0.20+fixes14472/programs/mythfrontend/globalsettings.cpp
--- blah/programs/mythfrontend/globalsettings.cpp	(revision 14472)
+++ blah/programs/mythfrontend/globalsettings.cpp	(working copy)
@@ -3638,7 +3638,7 @@
 
     VerticalConfigurationGroup* screen = new VerticalConfigurationGroup(false);
     screen->setLabel(QObject::tr("Screen settings"));
-    if (GetNumberOfXineramaScreens())
+    if (GetNumberOfXineramaScreens() > 1)
     {
         screen->addChild(XineramaScreen());
         screen->addChild(XineramaMonitorAspectRatio());
diff -urNad mythtv-0.20+fixes14472~/programs/mythfrontend/networkcontrol.cpp mythtv-0.20+fixes14472/programs/mythfrontend/networkcontrol.cpp
--- blah/programs/mythfrontend/networkcontrol.cpp	(revision 14472)
+++ blah/programs/mythfrontend/networkcontrol.cpp	(working copy)
@@ -192,7 +192,7 @@
     if (result != "")
     {
         nrLock.lock();
-        networkControlReplies.push_back(result);
+        networkControlReplies.push_back(QDeepCopy<QString>(result));
         nrLock.unlock();
 
         notifyDataAvailable();
@@ -227,8 +227,13 @@
     client = s;
     cs = os;
 
+    ncLock.lock();
     networkControlCommands.clear();
+    ncLock.unlock();
+
+    nrLock.lock();
     networkControlReplies.clear();
+    nrLock.unlock();
 
     welcomeStr = "MythFrontend Network Control\r\n";
     if (closedOldConn)
@@ -263,12 +268,15 @@
         lineIn.replace(QRegExp("[\r\n]"), "");
         lineIn.replace(QRegExp("^\\s"), "");
 
+        if (lineIn.isEmpty())
+            continue;
+
         tokens = QStringList::split(" ", lineIn);
-        if (tokens[0].lower() != "key")
+        if (tokens.size() && (tokens[0].lower() != "key"))
             lineIn = lineIn.lower();
 
         ncLock.lock();
-        networkControlCommands.push_back(lineIn);
+        networkControlCommands.push_back(QDeepCopy<QString>(lineIn));
         ncLock.unlock();
     }
 }
@@ -723,7 +731,7 @@
             for (unsigned int i = 3; i < tokens.size(); i++)
                 response += QString(" ") + tokens[i];
             nrLock.lock();
-            networkControlReplies.push_back(response);
+            networkControlReplies.push_back(QDeepCopy<QString>(response));
             nrLock.unlock();
 
             notifyDataAvailable();
diff -urNad mythtv-0.20+fixes14472~/programs/mythtranscode/replex/replex.pro mythtv-0.20+fixes14472/programs/mythtranscode/replex/replex.pro
--- blah/programs/mythtranscode/replex/replex.pro	(revision 14472)
+++ blah/programs/mythtranscode/replex/replex.pro	(working copy)
@@ -20,8 +20,8 @@
 INCLUDEPATH += ../../../libs/libavutil  ../../../libs/libmythmpeg2
 
 LIBS += -lmythavcodec-$$LIBVERSION -lmythavformat-$$LIBVERSION -lmythavutil-$$LIBVERSION
+LIBS += -L../../../libs/libavutil -L../../../libs/libavcodec -L../../../libs/libavformat
 LIBS += $$EXTRA_LIBS
-LIBS += -L../../../libs/libavutil -L../../../libs/libavcodec -L../../../libs/libavformat
 
 isEmpty(QMAKE_EXTENSION_SHLIB) {
   QMAKE_EXTENSION_SHLIB=so
diff -urNad mythtv-0.20+fixes14472~/programs/mythtranscode/transcode.cpp mythtv-0.20+fixes14472/programs/mythtranscode/transcode.cpp
--- blah/programs/mythtranscode/transcode.cpp	(revision 14472)
+++ blah/programs/mythtranscode/transcode.cpp	(working copy)
@@ -593,7 +593,7 @@
     VideoOutput *videoOutput = nvp->getVideoOutput();
     bool is_key = 0;
     bool first_loop = true;
-    unsigned char *newFrame = new unsigned char[newWidth * newHeight * 3 / 2];
+    unsigned char *newFrame = new unsigned char[frame.size];
 
     frame.buf = newFrame;
     AVPicture imageIn, imageOut;
@@ -782,6 +782,7 @@
                 }
                 else
                 {
+                    frame.buf = newFrame;
                     avpicture_fill(&imageIn, lastDecode->buf, PIX_FMT_YUV420P,
                                    video_width, video_height);
                     avpicture_fill(&imageOut, frame.buf, PIX_FMT_YUV420P,
@@ -828,6 +829,7 @@
             }
             else
             {
+                frame.buf = newFrame;
                 avpicture_fill(&imageIn, lastDecode->buf, PIX_FMT_YUV420P,
                                video_width, video_height);
                 avpicture_fill(&imageOut, frame.buf, PIX_FMT_YUV420P,
diff -urNad mythtv-0.20+fixes14472~/programs/channel_changers/6200ch.c mythtv-0.20+fixes14472/programs/channel_changers/6200ch.c
--- blah/contrib/channel_changers/6200ch.c	(revision 14472)
+++ blah/contrib/channel_changers/6200ch.c	(working copy)
@@ -25,10 +25,28 @@
 #include <stdio.h>
 #include <errno.h>
 #include <stdlib.h>
+#include <limits.h> //for LLONG_MAX
+#include <getopt.h>
 #include <unistd.h> // for usleep
 
-// Motorola DCT-6200 IDs
-// Note: there are at least eleven different vendor IDs for the 6200
+// Vendor and Model IDs.
+// NOTE: Some Models have more than one possible Vendor ID
+// WARNING: Please update firewiredevice.cpp when adding to this list.
+
+#define DCH3200_VENDOR_ID1 0x00001c11
+#define DCH3200_MODEL_ID1  0x0000d330
+
+#define DCT3412_VENDOR_ID1 0x0000159a
+#define DCT3412_MODEL_ID1  0x000034cb
+
+#define DCT3416_VENDOR_ID1 0x000016b5
+#define DCT3416_VENDOR_ID2 0x00001bdd
+#define DCT3416_MODEL_ID1  0x0000346b
+#define DCT3416_MODEL_ID2  0x0000b630
+
+#define DCT5100_VENDOR_ID1 0x000017ee
+#define DCT5100_MODEL_ID1  0x0000620a
+
 #define DCT6200_VENDOR_ID1 0x00000ce5
 #define DCT6200_VENDOR_ID2 0x00000e5c
 #define DCT6200_VENDOR_ID3 0x00001225
@@ -40,13 +58,23 @@
 #define DCT6200_VENDOR_ID9 0x000014e8
 #define DCT6200_VENDOR_ID10 0x000016b5
 #define DCT6200_VENDOR_ID11 0x00001371
+#define DCT6200_VENDOR_ID12 0x000019a6
+#define DCT6200_VENDOR_ID13 0x00001aad
+#define DCT6200_VENDOR_ID14 0x00000b06
 #define DCT6200_SPEC_ID    0x00005068
 #define DCT6200_SW_VERSION 0x00010101
 #define DCT6200_MODEL_ID1  0x0000620a
 #define DCT6200_MODEL_ID2  0x00006200
+
 #define DCT6412_VENDOR_ID1 0x00000f9f
+#define DCT6412_VENDOR_ID2 0x0000152f
 #define DCT6412_MODEL_ID1  0x000064ca
+#define DCT6412_MODEL_ID2  0x000064cb
 
+#define DCT6416_VENDOR_ID1 0x000017ee
+#define DCT6416_VENDOR_ID2 0x00001a66 
+#define DCT6416_MODEL_ID1  0x0000646b
+
 #define AVC1394_SUBUNIT_TYPE_6200 (9 << 19)  /* uses a reserved subunit type */ 
 
 #define AVC1394_6200_COMMAND_CHANNEL 0x000007C00   /* 6200 subunit command */
@@ -161,8 +189,13 @@
       if (verbose) 
          printf("node %d: vendor_id = 0x%08x model_id = 0x%08x\n", 
                  i, dir.vendor_id, dir.model_id); 
-		
-      if ( ((dir.vendor_id == DCT6200_VENDOR_ID1) || 
+
+      if ( ((dir.vendor_id == DCH3200_VENDOR_ID1) ||
+            (dir.vendor_id == DCT3412_VENDOR_ID1) || 
+            (dir.vendor_id == DCT3416_VENDOR_ID1) || 
+            (dir.vendor_id == DCT3416_VENDOR_ID2) || 
+            (dir.vendor_id == DCT5100_VENDOR_ID1) ||
+            (dir.vendor_id == DCT6200_VENDOR_ID1) ||
             (dir.vendor_id == DCT6200_VENDOR_ID2) ||
             (dir.vendor_id == DCT6200_VENDOR_ID3) ||
             (dir.vendor_id == DCT6200_VENDOR_ID4) ||
@@ -173,10 +206,23 @@
             (dir.vendor_id == DCT6200_VENDOR_ID9) ||
             (dir.vendor_id == DCT6200_VENDOR_ID10) ||
             (dir.vendor_id == DCT6200_VENDOR_ID11) ||
-            (dir.vendor_id == DCT6412_VENDOR_ID1)) &&
-           ((dir.model_id == DCT6200_MODEL_ID1) ||
+            (dir.vendor_id == DCT6200_VENDOR_ID12) ||
+            (dir.vendor_id == DCT6200_VENDOR_ID13) ||
+            (dir.vendor_id == DCT6200_VENDOR_ID14) ||
+            (dir.vendor_id == DCT6412_VENDOR_ID1) ||
+            (dir.vendor_id == DCT6412_VENDOR_ID2) ||
+            (dir.vendor_id == DCT6416_VENDOR_ID1) || 
+            (dir.vendor_id == DCT6416_VENDOR_ID2)) &&
+           ((dir.model_id == DCH3200_MODEL_ID1) ||
+            (dir.model_id == DCT3412_MODEL_ID1) ||
+            (dir.model_id == DCT3416_MODEL_ID1) ||
+            (dir.model_id == DCT3416_MODEL_ID2) ||
+            (dir.model_id == DCT5100_MODEL_ID1) ||
+            (dir.model_id == DCT6200_MODEL_ID1) ||
             (dir.model_id == DCT6200_MODEL_ID2) ||
-            (dir.model_id == DCT6412_MODEL_ID1)) ) {
+            (dir.model_id == DCT6412_MODEL_ID1) ||
+            (dir.model_id == DCT6412_MODEL_ID2) ||
+            (dir.model_id == DCT6416_MODEL_ID1)) ) {
             if (dir.unit_spec_id != DCT6200_SPEC_ID)
                fprintf(stderr, "Warning: Unit Spec ID different.\n");
             if (dir.unit_sw_version != DCT6200_SW_VERSION)
diff -urNad mythtv-0.20+fixes14472~/programs/channel_changers/sa3250ch.c mythtv-0.20+fixes14472/programs/channel_changers/sa3250ch.c
--- blah/contrib/channel_changers/sa3250ch.c	(revision 14472)
+++ blah/contrib/channel_changers/sa3250ch.c	(working copy)
@@ -31,12 +31,16 @@
 #include <string.h>
 
 /* SA3250HD IDs */
-#define SA3250HD_VENDOR_ID1     0x000011e6
-#define SA3250HD_VENDOR_ID2     0x000014f8
-#define SA3250HD_VENDOR_ID3     0x00001692
+#define SA_VENDOR_ID1           0x000011e6
+#define SA_VENDOR_ID2           0x000014f8
+#define SA_VENDOR_ID3           0x00001692
+#define SA_VENDOR_ID4           0x00001947
+#define SA_VENDOR_ID5           0x00000f21
+#define SA_VENDOR_ID6           0x00001ac3
+#define SA_VENDOR_ID7           0x00000a73
 #define SA3250HD_MODEL_ID1      0x00000be0
-#define SA4200HD_VENDOR_ID1     0x000014f8
 #define SA4200HD_MODEL_ID1      0x00001072
+#define SA4250HDC_MODEL_ID1     0x000010cc
 
 #define AVC1394_SA3250_COMMAND_CHANNEL 0x000007c00   /* subunit command */
 #define AVC1394_SA3250_OPERAND_KEY_PRESS 0xe7
@@ -53,7 +57,8 @@
 {
    fprintf(stderr, "Usage: sa3250ch [-v] [-s] <channel_num>\n");
    fprintf(stderr, "  -v : Verbose Mode\n");
-   fprintf(stderr, "  -s : Send command as single digit (for SA4200 and some SA3250s)\n");
+   fprintf(stderr, "  -s : Send command as single digit "
+           "(for SA4200 and some SA3250s and SA4200HD's)\n");
    exit(1);
 }
 
@@ -123,19 +128,36 @@
          printf("node %d: vendor_id = 0x%08x model_id = 0x%08x\n", 
                  i, dir.vendor_id, dir.model_id); 
 		
-      if ( ((dir.vendor_id == SA4200HD_VENDOR_ID1) &&
-	    (dir.model_id == SA4200HD_MODEL_ID1))  ||
-          (((dir.vendor_id == SA3250HD_VENDOR_ID1) ||
-            (dir.vendor_id == SA3250HD_VENDOR_ID2) ||
-            (dir.vendor_id == SA3250HD_VENDOR_ID3)) &&
-            (dir.model_id == SA3250HD_MODEL_ID1))) {
-            device = i;
-            break;
+      /* WARNING: Please update firewiredevice.cpp when adding to this list. */
+      if (((dir.vendor_id == SA_VENDOR_ID1) ||
+           (dir.vendor_id == SA_VENDOR_ID2) ||
+           (dir.vendor_id == SA_VENDOR_ID3) ||
+           (dir.vendor_id == SA_VENDOR_ID4) ||
+           (dir.vendor_id == SA_VENDOR_ID5) ||
+           (dir.vendor_id == SA_VENDOR_ID6) ||
+           (dir.vendor_id == SA_VENDOR_ID7)) &&
+          ((dir.model_id == SA3250HD_MODEL_ID1)  ||
+           (dir.model_id == SA4200HD_MODEL_ID1)  ||
+           (dir.model_id == SA4250HDC_MODEL_ID1)))
+      {
+          if (dir.model_id == SA4250HDC_MODEL_ID1)
+          {
+              fprintf(stderr, "Ignoring SA 4250HDC on node %d -- "
+                      "sorry not supported yet\n", i);
+          }
+          else
+          {
+              device = i;
+              break;
+          }
       }
    }
-    
-   if (device == -1) {
-        fprintf(stderr, "Could not find SA3250HD on the 1394 bus.\n");
+
+   if (device == -1)
+   {
+        fprintf(stderr, "Could not find SA3250HD or SA4200HD "
+                "on the IEEE 1394 bus.\n");
+
         raw1394_destroy_handle(handle);
         exit(1);
    }
